File: web/src/utils/ma.js
/*
 * @Description: 计算 MA 均线数据
 * @Date: 2024-01-07 19:05:47
 */

/*
 * 计算 MA 均线数据
 * @param {number} dayCount 几日的均线
 * @param {array} data 数据
 */
export function calculateMA(dayCount, data) {
  const result = []
  let len = data.length
  for (let i = 0; i < len; i++) {
    if (i < dayCount) {
      result.push('-')
      continue
    }
    let sum = 0
    for (let j = 0; j < dayCount; j++) {
      sum += parseFloat(data[i - j][1])
    }
    // 保留2位小数
    result.push((sum / dayCount).toFixed(2))
  }
  return result
}

File: server-django/stock_manage/utils/MA.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Description: 计算 MA 均线数据
# @Author   : jinhe Cen
# @Time     : 2024/2/25 12:17
# @File     : MA.py


# 计算 MA 均线数据
def calculateMA(dayCount, data, field=1):
    """
    :param dayCount: 几日均线
    :param data: 数据
    :param field: 计算数据的哪个字段,如: 收盘(close)
    :return: 均线数据
    """
    result = []
    length = len(data)
    for i in range(length):
        if i < dayCount:
            result.append('-')
            continue
        total = 0
        # 取dayCount的合计
        for j in range(dayCount):
            total += float(data[i - j][field])
        # 取dayCount的平均值,保留2位小数
        element = total / dayCount
        result.append('{:.2f}'.format(element))
    return result
